numba/misc/gdb_print_extension.py

```diff
@@ -0,0 +1,186 @@
+                dtype_str, ndim_str, order_str = arr_info
+                rstr = 'Record\\((.*\\[.*\\]);([0-9]+);(True|False)'
+                fields, balign, is_aligned = re.match(rstr, dtype_str).groups()
+                field_dts = fields.split(',')
+                struct_entries = []
+                for f in field_dts:
+                    name, stuff = f.split('[')
+                    dt_as_str = stuff.split(';')[0].split('=')[1]
+                    if "unichr" in dt_as_str:
+                        raise ValueError
+                    else:
+                        dtype = np.dtype(dt_as_str)
+                    struct_entries.append((name, dtype))
+                    # The dtype is actually a record of some sort
+                dtype_str = struct_entries
+            else:  # simple type
+                matcher = re.compile(r"array\((.*),(.*),(.*)\)\ \(.*")
+                arr_info = [x.strip() for x in matcher.match(ty_str).groups()]
+                dtype_str, ndim_str, order_str = arr_info
+
+            def dwarr2inttuple(dwarr):
+                """Converts a gdb handle to a dwarf array to a tuple of ints"""
+                fields = dwarr.type.fields()
+                lo, hi = fields[0].type.range()
+                return tuple([int(dwarr[x]) for x in range(lo, hi + 1)])
+
+            # shape/strides extraction
+            shape = dwarr2inttuple(rshp)
+            strides = dwarr2inttuple(rstrides)
+
+            # if data is not NULL
+            if data != NULL:
+                if HAVE_NUMPY:
+                    # The data extent in bytes is:
+                    # sum(shape * strides)
+                    # get the data, then wire to as_strided
+                    shp_arr = np.array([x - 1 for x in shape])
+                    strd_arr = np.array(strides)
+                    extent = np.sum(shp_arr * strd_arr)
+                    extent += int(itemsize)
+                    dtype_clazz = np.dtype(dtype_str, align=is_aligned)
+                    dtype = dtype_clazz  # .type
+                    this_proc = gdb.selected_inferior()
+                    mem = this_proc.read_memory(int(data), extent)
+                    arr_data = np.frombuffer(mem, dtype=dtype)
+                    new_arr = np.lib.stride_tricks.as_strided(arr_data,
+                                                              shape=shape,
+                                                              strides=strides,)
+                    return '\n' + str(new_arr)
+                # Catch all for no NumPy
+                return "array([...], dtype=%s, shape=%s)" % (dtype_str, shape)
+            else:
+                # Not yet initialized or NULLed out data
+                buf = list(["NULL/Uninitialized"])
+                return "array([" + ', '.join(buf) + "]" + ")"
+        except Exception as e:
+            return 'array[Exception: Failed to parse. %s]' % e
```

Comment:
This function is complicated enough that I really want it tested regularly even without gdb. We should be able to mock the gdb interface. The parts like parsing the `Record` type description should perhaps be part of the `Record` type as well.

Reply:
Testing the function is addressed in https://github.com/numba/numba/pull/7617/commits/3f146ba25ed51f6b43369384cb7387bcf7dd05f9